java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java

```diff
@@ -0,0 +1,123 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.hudi.reader;
+
+import com.starrocks.jni.connector.ColumnValue;
+import org.apache.hadoop.hive.serde2.objectinspector.ListObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.MapObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.StructField;
+import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
+
+import java.util.List;
+import java.util.Map;
+
+public class HudiColumnValue implements ColumnValue {
+    private Object fieldData;
+    private ObjectInspector fieldInspector;
+
+    HudiColumnValue(ObjectInspector fieldInspector, Object fieldData) {
+        this.fieldInspector = fieldInspector;
+        this.fieldData = fieldData;
+    }
+
+    private Object inspectObject() {
+        return ((PrimitiveObjectInspector) fieldInspector).getPrimitiveJavaObject(fieldData);
+    }
+
+    @Override
+    public boolean getBoolean() {
+        return (boolean) inspectObject();
+    }
+
+    @Override
+    public short getShort() {
+        return (short) inspectObject();
+    }
+
+    @Override
+    public int getInt() {
+        return (int) inspectObject();
+    }
+
+    @Override
+    public float getFloat() {
+        return (float) inspectObject();
+    }
+
+    @Override
+    public long getLong() {
+        return (long) inspectObject();
+    }
+
+    @Override
+    public double getDouble() {
+        return (double) inspectObject();
+    }
+
+    @Override
+    public String getString() {
+        return inspectObject().toString();
+    }
+
+    @Override
+    public void unpackArray(List<ColumnValue> values) {
+        ListObjectInspector inspector = (ListObjectInspector) fieldInspector;
+        List items = inspector.getList(fieldData);
+        ObjectInspector itemInspector = inspector.getListElementObjectInspector();
+        for (Object item : items) {
+            HudiColumnValue cv = null;
+            if (item != null) {
+                cv = new HudiColumnValue(itemInspector, item);
+            }
+            values.add(cv);
+        }
+    }
+
+    @Override
+    public void unpackMap(List<ColumnValue> keys, List<ColumnValue> values) {
+        MapObjectInspector inspector = (MapObjectInspector) fieldInspector;
+        ObjectInspector keyObjectInspector = inspector.getMapKeyObjectInspector();
+        ObjectInspector valueObjectInspector = inspector.getMapValueObjectInspector();
+        for (Map.Entry kv : inspector.getMap(fieldData).entrySet()) {
+            HudiColumnValue cv0 = null;
+            HudiColumnValue cv1 = null;
+            if (kv.getKey() != null) {
```

Comment:
Need to unpack recursively? map value maybe struct or struct

Reply:
Yes, need to unpack recursively. In theory map's key usually is int/string, but here we can handle map's key in a more general way. No harm at all.